### ISSUE TITLE
[2.5.9] LIST (SPECIAL-USE) different on frontends compared to backend

### ISSUE THREAD

**Issue by MASHtm** [cyrusimap/cyrus-imapd#45]
First I thought this is the same as #11, but i think it's closer to #41. So I file a new issue.

if I do
```
A0005 LIST (SPECIAL-USE) "" "*" RETURN (SUBSCRIBED)
```
my backend answers with
```
* LIST (\Subscribed \HasNoChildren \Drafts) "." INBOX.Drafts
* LIST (\Subscribed \HasNoChildren \Junk) "." INBOX.Junk
* LIST (\Subscribed \HasNoChildren \Sent) "." INBOX.Sent
* LIST (\Subscribed \HasNoChildren \Trash) "." INBOX.Trash
A0005 OK Completed (0.070 secs 115 calls)
```
should'nt there be a default "\Inbox" entry for INBOX as well?

my frontends on the other hand show
```
A0005 LIST (SPECIAL-USE) "" "*" RETURN (SUBSCRIBED)
* LIST (\Subscribed \HasChildren) "." "INBOX"
* LIST (\Subscribed \HasChildren) "." "INBOX.Archiv"
...
* LIST (\Subscribed) "." "INBOX.Archiv.2016"
* LIST (\Subscribed) "." "INBOX.Drafts"
* LIST (\Subscribed) "." "INBOX.Junk"
* LIST (\Subscribed) "." "INBOX.Sent"
* LIST (\Subscribed) "." "INBOX.Templates"
* LIST (\Subscribed) "." "INBOX.Trash"
...
* LIST (\Subscribed \HasChildren) "." "user.dmaxx"
...
* LIST (\Subscribed \HasChildren) "." "user.maxxx"
...
A0005 OK Completed (0.010 secs)
```
no special-use at all and all folders I've subscribed.

**Comment by elliefm** [comment]
I'm working on a Cassandane test to reproduce this.

Testing against cyrus-imapd-2.5 at ab4f903 (a couple of commits ago), I see this on the frontend:

```
C: 3 create INBOX.Drafts
S: 3 OK Completed
C: 4 subscribe INBOX.Drafts
S: 4 OK Completed
C: 5 setmetadata INBOX.Drafts ("/private/specialuse" "\\Drafts")
S: 5 OK Completed
C: 6 create INBOX.Junk
S: 6 OK Completed
C: 7 subscribe INBOX.Junk
S: 7 OK Completed
C: 8 setmetadata INBOX.Junk ("/private/specialuse" "\\Junk")
S: 8 OK Completed
C: 9 create INBOX.Sent
S: 9 OK Completed
C: 10 subscribe INBOX.Sent
S: 10 OK Completed
C: 11 setmetadata INBOX.Sent ("/private/specialuse" "\\Sent")
S: 11 OK Completed
C: 12 create INBOX.Trash
S: 12 OK Completed
C: 13 subscribe INBOX.Trash
S: 13 OK Completed
C: 14 setmetadata INBOX.Trash ("/private/specialuse" "\\Trash")
S: 14 OK Completed
C: 15 list (SPECIAL-USE) "" "*" RETURN (SUBSCRIBED)
S: * LIST (\Subscribed) "." "INBOX.Drafts"
S: * LIST (\Subscribed) "." "INBOX.Junk"
S: * LIST (\Subscribed) "." "INBOX.Sent"
S: * LIST (\Subscribed) "." "INBOX.Trash"
S: 15 OK Completed (0.000 secs)
```

And then this on the backend:

```
C: 16 list (SPECIAL-USE) "" "*" RETURN (SUBSCRIBED)
S: * LIST (\Subscribed) "." "INBOX.Drafts"
S: * LIST (\Subscribed) "." "INBOX.Junk"
S: * LIST (\Subscribed) "." "INBOX.Sent"
S: * LIST (\Subscribed) "." "INBOX.Trash"
S: 16 OK Completed (0.000 secs)
```

**Comment by elliefm** [comment]
Testing against current cyrus-imapd-2.5 (with the addition of 5dd805e and 36937f3), on the frontend I get:

```
C: 3 create INBOX.Drafts
S: 3 OK Completed
C: 4 subscribe INBOX.Drafts
S: 4 OK Completed
C: 5 setmetadata INBOX.Drafts ("/private/specialuse" "\\Drafts")
S: 5 OK Completed
C: 6 create INBOX.Junk
S: 6 OK Completed
C: 7 subscribe INBOX.Junk
S: 7 OK Completed
C: 8 setmetadata INBOX.Junk ("/private/specialuse" "\\Junk")
S: 8 OK Completed
C: 9 create INBOX.Sent
S: 9 OK Completed
C: 10 subscribe INBOX.Sent
S: 10 OK Completed
C: 11 setmetadata INBOX.Sent ("/private/specialuse" "\\Sent")
S: 11 OK Completed
C: 12 create INBOX.Trash
S: 12 OK Completed
C: 13 subscribe INBOX.Trash
S: 13 OK Completed
C: 14 setmetadata INBOX.Trash ("/private/specialuse" "\\Trash")
S: 14 OK Completed
C: 15 list (SPECIAL-USE) "" "*" RETURN (SUBSCRIBED)
S: * LIST (\Subscribed \HasNoChildren) "." "INBOX.Drafts"
S: * LIST (\Subscribed \HasNoChildren) "." "INBOX.Junk"
S: * LIST (\Subscribed \HasNoChildren) "." "INBOX.Sent"
S: * LIST (\Subscribed \HasNoChildren) "." "INBOX.Trash"
S: 15 OK Completed (0.000 secs)
```

And then on the backend I get:

```
C: 16 list (SPECIAL-USE) "" "*" RETURN (SUBSCRIBED)
S: * LIST (\Subscribed \HasNoChildren) "." "INBOX.Drafts"
S: * LIST (\Subscribed \HasNoChildren) "." "INBOX.Junk"
S: * LIST (\Subscribed \HasNoChildren) "." "INBOX.Sent"
S: * LIST (\Subscribed \HasNoChildren) "." "INBOX.Trash"
S: 16 OK Completed (0.000 secs)
```

Which is interesting because none of these cases are returning the special-use flags, but the results _are_ all consistent between frontend and backend.

**Comment by elliefm** [comment]
If I explicitly ask for return SPECIAL-USE, then I get it (also latest cyrus-imapd-2.5):

Frontend:

```
[snip the same create/subscribe/setmetadata setup]
C: 15 list (SPECIAL-USE) "" "*" RETURN (SUBSCRIBED SPECIAL-USE)
S: * LIST (\Subscribed \HasNoChildren \Drafts) "." "INBOX.Drafts"
S: * LIST (\Subscribed \HasNoChildren \Junk) "." "INBOX.Junk"
S: * LIST (\Subscribed \HasNoChildren \Sent) "." "INBOX.Sent"
S: * LIST (\Subscribed \HasNoChildren \Trash) "." "INBOX.Trash"
S: 15 OK Completed (0.000 secs)
```

Backend:

```
C: 16 list (SPECIAL-USE) "" "*" RETURN (SUBSCRIBED SPECIAL-USE)
S: * LIST (\Subscribed \HasNoChildren \Drafts) "." "INBOX.Drafts"
S: * LIST (\Subscribed \HasNoChildren \Junk) "." "INBOX.Junk"
S: * LIST (\Subscribed \HasNoChildren \Sent) "." "INBOX.Sent"
S: * LIST (\Subscribed \HasNoChildren \Trash) "." "INBOX.Trash"
S: 16 OK Completed (0.000 secs)
```

I can't see any commits since 2.5.9 that seem like might make any difference here (other than the two I called out above), so it's weird that we're seeing such distinctly different behaviours?

**Comment by elliefm** [comment]
Cassandane test is here: https://github.com/cyrusimap/cassandane/commit/62b61f637dbf62a3f92f6cdd615df6d6289326f5

**Comment by MASHtm** [comment]
is it possible that this is caused by migrating from 2.4? I migrated my backends first and set the metadata directly on the backends using 2.5.8 while mupdate and frontends still used 2.4.

My RPMs are bulit from tarball downloaded from 
ftp://ftp.cyrusimap.org/cyrus-imapd/cyrus-imapd-2.5.9.tar.gz

**Comment by elliefm** [comment]
Do you have the `specialusealways` option set in imapd.conf, I wonder?  That might explain why you get the special use flags returned (sometimes) whereas I only get them when I ask for them explicitly.

As for the inconsistency -- it might be a side effect of migration from 2.4, but I don't know what to look for to verify if that's the case (much less how to fix it)

Actually, I've just caught a detail in your comment:

> I [...] set the metadata directly on the backends

In my tests I'm setting it on the front end, then interrogating the front end, then interrogating the back end.  I wonder what happens if I set it directly on the back end, whether the front end can see it...

**Comment by elliefm** [comment]
... nope, doing the create/subscribe/setmetadata step on the backend instead of the frontend makes no difference for me, both the frontend and the backend still subsequently respond with identical and valid-looking list responses.  Curious.

**Comment by MASHtm** [comment]
I use
```
specialusealways: 0
```
on all my hosts

**Comment by elliefm** [comment]
> should'nt there be a default "\Inbox" entry for INBOX as well?

My vague understanding is that this is an XLIST thing, not a SPECIAL-USE thing.  RFC 6154 doesn't define \Inbox, and (at least on master, which I happen to be looking at just now) our support of \Inbox depends specifically on the command being XLIST (leading me to this understanding).

**Comment by elliefm** [comment]
I've just run my test with 2.5.9 and I get the same results as in my first comment -- so the difference in behaviour isn't due to newer commits in git.  

Which I guess means there must be some aspect of your environment that's triggering this behaviour, which my minimal test isn't replicating yet.

Do your frontend and backend report the same results for `getmetadata folder "/private/specialuse"` for these folders?

**Comment by elliefm** [comment]
I've just had a thought: my test runs both frontend and backend on the same machine, with only a bunch of different port numbers/config directories between them, so maybe all the "localhost" / "127.0.0.1" in my configuration is accidentally bypassing the murder code...

Having a look at `list_data_remote()` in `imap/imapd.c` looks a lot like it would do what you're seeing -- so maybe my murder doesn't really think it's a murder, and my "frontend" and "backend" are both kind of the same thing...

There's a fix on master for `list_data_remote()` to make it proxy extended list commands completely (instead of the fudged way it does it in 2.5), so the solution here may be to just backport that.

I'll see dig deeper into my test murder setup and, if it's doing what I suspect it's doing, see if I can make it work properly, in which case I think I should then be able to reproduce this.

**Comment by MASHtm** [comment]
`Do your frontend and backend report the same results for getmetadata folder "/private/specialuse" for these folders?`

Yes, both give the same answer to getmetadata. See #41

**Comment by elliefm** [comment]
I must've been asleep at the wheel a couple of days ago, I can reproduce this exactly as you describe now.

My tests were a) testing with an account that only had special-use folders, so I wasn't seeing any unexpected extra stuff in the output, and b) listing against the frontend twice instead of the frontend once and the backend once.  So all the previous test results in my earlier comments are utter rubbish. Doh!

Anyway, this behaviour looks very much like what 67bac5df fixes on master, so I'll fix up my tests and then see how that backports...

**Comment by elliefm** [comment]
Backported here: https://github.com/cyrusimap/cyrus-imapd/compare/cyrus-imapd-2.5...elliefm:v25/issue45

This fixes the issue for me, I wonder whether it helps with #41 too...

**Comment by MASHtm** [comment]
have built a new version with it (including the fix for my selflart;-) )... will try and report.

**Comment by MASHtm** [comment]
a quick test ... fixes this bug, but not #41 
```
. list (special-use) "" "*" return (subscribed)
* LIST (\Subscribed \HasNoChildren \Drafts) "." "INBOX.Drafts"
* LIST (\Subscribed \HasNoChildren \Junk) "." "INBOX.Junk"
* LIST (\Subscribed \HasNoChildren \Sent) "." "INBOX.Sent"
* LIST (\Subscribed \HasNoChildren \Trash) "." "INBOX.Trash"
. OK Completed (0.000 secs)
```
on both backend and frontend, but xlist on frontend doesn't contain special-use flags..

**Comment by elliefm** [comment]
Fix is now on cyrus-imapd-2.5 branch and will be included in next release
